File: strategy/marvin/agent/defender.lua
local Base = require "agent/base/agent"
local Defender = Class("Agent.Defender", Base)

local World = require "../base/world"

local Default = require "agent/defender/default"
local HandleBall = require "agent/defender/handleball"
local ManMark = require "agent/defender/manmark"
local ZoneDefense = require "agent/defender/zonedefense"
local Penalty = require "agent/defender/penalty"

local Messaging = require "control/messaging"
local Move = require "moves/moveBehavior"


Defender._behaviors = {
	Penalty,
	Move,
	HandleBall,
	ManMark,
	ZoneDefense,
	Default
}

function Defender:_run()
	self._activeBehavior._send.defenderFlag("all")
end

function Defender.takeRobot(robots)
	for _, robot in ipairs(robots) do
		if robot.isVisible then
			return robot
		end
	end
end

function Defender:keepRobot()
	return self._robot.isVisible and self._robot ~= World.FriendlyKeeper and not self._robot.userControl
end

-- worse rating if robot if farther away from own goal
function Defender:rateRobot()
	if self._activeBehavior and self._activeBehavior:forceKeepingInPool() then
		return math.huge
	end
	return -World.Geometry.FriendlyGoal:distanceTo(self._robot.pos)
end

return Defender

File: strategy/typescript/glados/agent/defender.ts
import { FriendlyRobot } from "base/robot";
import * as World from "base/world";

import { Agent } from "glados/agent/base/agent";
import { BehaviorConstructor } from "glados/agent/base/behavior";
import { CenterBack } from "glados/agent/defender/centerback";
import { Default } from "glados/agent/defender/default";
import { HandleBall } from "glados/agent/defender/handleball";
import { ManMark } from "glados/agent/defender/manmark";
import { PassTarget } from "glados/agent/defender/passtarget";
import { PenaltyPassiveDefenderOffense } from "glados/agent/defender/penaltypassivedefenderoffense";
import { Piggy } from "glados/agent/defender/piggy";
import { ZoneDefense } from "glados/agent/defender/zonedefense";
import { BallEscort } from "glados/agent/shared/ballescort";
import { BreakPass } from "glados/agent/shared/breakpass";
import { PenaltyPassiveDefense } from "glados/agent/shared/penaltypassivedefense";
import { MessageType } from "glados/control/messaging";


export class Defender extends Agent {

	public getBehaviors(): BehaviorConstructor[] {
		return [
			PenaltyPassiveDefenderOffense,
			PenaltyPassiveDefense,
			PassTarget,
			BallEscort,
			HandleBall,
			ManMark,
			CenterBack,
			BreakPass,
			Piggy,
			ZoneDefense,
			Default
		];
	}

	public _run() {
		this._activeBehavior!._messaging.sendBroadcast(MessageType.defenderFlag);
	}

	public static takeRobot(robots: FriendlyRobot[]): FriendlyRobot | undefined {
		for (let robot of robots) {
			if (robot.isVisible) {
				return robot;
			}
		}
		return undefined;
	}

	public keepRobot(): boolean {
		return this._robot.isVisible && this._robot !== World.FriendlyKeeper && this._robot.userControl == undefined;
	}

	// worse rating if robot if farther away from own goal
	public rateRobot(): number {
		if (this._activeBehavior != undefined && this._activeBehavior.forceKeepingInPool()) {
			return Infinity;
		}
		let passInfoTable = this._messaging.receiveSingleSender(MessageType.passInfo)[1];
		if (passInfoTable) {
			for (let pass of passInfoTable) {
				if (pass.target === this._robot) {
					// We are a very bad defender if we want to accept this pass :O
					return -Infinity;
				}
			}
		}
		return -World.Geometry.FriendlyGoal.distanceTo(this._robot.pos);
	}
}
